core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -416,78 +351,116 @@ private long getCurrentBlockId() {
+    // We should really only close a cache stream here. This check is to verify this.
+    Preconditions.checkState(mCurrentCacheStream == null || mCurrentCacheStream.remaining() == 0);
+    closeOrCancelCacheStream();
+    Preconditions.checkState(mCurrentCacheStream == null);
+
+    if (blockId < 0) {
+      // End of file.
+      return;
+    }
+    Preconditions.checkNotNull(mCurrentBlockInStream);
+    if (!mShouldCache || mCurrentBlockInStream instanceof LocalBlockInStream) {
+      return;
+    }
+
+    // Unlike updateBlockInStream below, we never start a block cache stream if mPos is in the
+    // middle of a block.
+    if (mPos % mBlockSize != 0) {
+      return;
+    }
+
+    try {
+      WorkerNetAddress address = mLocationPolicy.getWorkerForNextBlock(
+          mContext.getAlluxioBlockStore().getWorkerInfoList(), getBlockSizeAllocation(mPos));
+      // Don't cache the block to somewhere that already has it.
+      // TODO(andrew,peis): Filter the workers provided to the location policy to not include
+      // workers which already contain the block. See ALLUXIO-1816.
+      if (mCurrentBlockInStream instanceof RemoteBlockInStream) {
+        WorkerNetAddress readAddress =
+            ((RemoteBlockInStream) mCurrentBlockInStream).getWorkerNetAddress();
+        // Try to avoid an RPC.
+        if (readAddress.equals(address)) {
+          return;
+        }
+        BlockInfo blockInfo = mContext.getAlluxioBlockStore().getInfo(blockId);
+        for (BlockLocation location : blockInfo.getLocations()) {
+          if (address.equals(location.getWorkerAddress())) {
+            return;
+          }
         }
-      } else {
-        mShouldCacheCurrentBlock = false;
       }
+      // If we reach here, we need to cache.
+      mCurrentCacheStream =
+          mContext.getAlluxioBlockStore().getOutStream(blockId, getBlockSize(mPos), address);
+    } catch (IOException e) {
+      handleCacheStreamIOException(e);
+    } catch (AlluxioException e) {
+      LOG.warn(BLOCK_ID_NOT_CACHED, blockId, e);
     }
   }
 
   /**
-   * Helper method to {@link #checkAndAdvanceBlockInStream()} and {@link #seekBlockInStream(long)}.
-   * The current {@link BlockInStream} will be closed and a new {@link BlockInStream} for the given
-   * blockId will be opened at position 0.
+   * Update {@link #mCurrentBlockInStream} to be in-sync with mPos's block. The new block
+   * stream created with be at position 0.
+   * This function is only called in {@link #updateStreams()}.
    *
-   * @param blockId blockId to set the {@link #mCurrentBlockInStream} to read
+   * @param blockId cached result of {@link #getCurrentBlockId()}
    * @throws IOException if the next {@link BlockInStream} cannot be obtained
    */
   private void updateBlockInStream(long blockId) throws IOException {
     if (mCurrentBlockInStream != null) {
       mCurrentBlockInStream.close();
+      mCurrentBlockInStream = null;
+    }
+    if (blockId < 0) {
```

Comment:
is this an expected case (and when will we see negative blockId if it expected?)?
if not expected, at least we should log error